// @ephox/mcagar/lib/main/ts/ephox/mcagar/api/LegacyUnit.js
import { Assertions, Logger, Step } from '@ephox/agar';
import { TinyDom } from './TinyDom';
var test = function (message, fn) {
    return function (initValue) {
        return Logger.t(message, Step.sync(function () {
            fn(initValue);
        }));
    };
};
var asyncTest = function (message, fn) {
    return function (initValue) {
        return Logger.t(message, Step.async(function (done, die) {
            fn(initValue, done, die);
        }));
    };
};
var createSuite = function () {
    var tests = [];
    return {
        test: function (message, fn) {
            tests.push(test(message, fn));
        },
        asyncTest: function (message, fn) {
            tests.push(asyncTest(message, fn));
        },
        toSteps: function (initValue) {
            return tests.map(function (test) {
                return test(initValue);
            });
        }
    };
};
var execCommand = function (editor, cmd, ui, value) {
    if (editor.editorCommands.hasCustomCommand(cmd)) {
        editor.execCommand(cmd, ui, value);
    }
};
var findContainer = function (editor, selector) {
    var container;
    if (typeof selector === 'string') {
        container = editor.dom.select(selector)[0];
    }
    else {
        container = selector;
    }
    if (container.firstChild) {
        container = container.firstChild;
    }
    return container;
};
var setSelection = function (editor, startSelector, startOffset, endSelector, endOffset) {
    var startContainer = findContainer(editor, startSelector);
    var endContainer = findContainer(editor, endSelector ? endSelector : startSelector);
    var rng = editor.dom.createRng();
    var setRange = function (container, offset, start) {
        offset = offset ? offset : 0;
        if (offset === 'after') {
            if (start) {
                rng.setStartAfter(container);
            }
            else {
                rng.setEndAfter(container);
            }
            return;
        }
        else if (offset === 'afterNextCharacter') {
            container = container.nextSibling;
            offset = 1;
        }
        if (start) {
            rng.setStart(container, offset);
        }
        else {
            rng.setEnd(container, offset);
        }
    };
    setRange(startContainer, startOffset, true);
    setRange(endContainer, endSelector ? endOffset : startOffset, false);
    editor.selection.setRng(rng);
};
var trimBrs = function (html) {
    return html.toLowerCase().replace(/<br[^>]*>|[\r\n]+/gi, '');
};
var equalDom = function (actual, expected, message) {
    Assertions.assertDomEq(typeof message !== 'undefined' ? message : 'Nodes are not equal', TinyDom.fromDom(expected), TinyDom.fromDom(actual));
};
var equal = function (actual, expected, message) {
    Assertions.assertEq(typeof message !== 'undefined' ? message : 'No message specified', expected, actual);
};
var strictEqual = equal;
var deepEqual = equal;
export { test, asyncTest, createSuite, execCommand, setSelection, trimBrs, equal, equalDom, strictEqual, deepEqual };
