// @ephox/bedrock-server/lib/main/ts/bedrock/server/Matchers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url = require("url");
const Cmp = require("../util/Cmp");
const Obj = require("../util/Obj");
exports.prefixMatch = (prefix) => {
    return (request) => {
        return request.url !== undefined && request.url.indexOf(prefix) === 0;
    };
};
exports.methodMatch = (method) => {
    return (request) => {
        return request.method === method.toUpperCase();
    };
};
exports.urlMatch = (url) => {
    return (request) => {
        return request.url === url;
    };
};
exports.pathMatch = (path) => {
    return (request) => {
        return url.parse(request.originalUrl).pathname === path;
    };
};
exports.headersMatch = (headers) => {
    const lcHeaders = Obj.toLowerCaseKeys(headers);
    return (request) => {
        return Cmp.hasAllOf(Obj.toLowerCaseKeys(request.headers), lcHeaders);
    };
};
exports.queryMatch = (query) => {
    return (request) => {
        const reqQuery = url.parse(request.originalUrl, true).query;
        return Cmp.hasAllOf(reqQuery, query);
    };
};
exports.jsonBodyMatch = (json) => {
    return (request) => {
        const data = request.body ? JSON.parse(request.body) : {};
        return Cmp.deepEq(json, data);
    };
};
