// @ephox/dom-globals/lib/main/ts/ephox/dom-globals/api/Main.js
var ANGLE_instanced_arrays$1 = window['ANGLE_instanced_arrays'];
var AbortController$1 = window['AbortController'];
var AbortSignal$1 = window['AbortSignal'];
var AnalyserNode$1 = window['AnalyserNode'];
var Animation$1 = window['Animation'];
var AnimationEvent$1 = window['AnimationEvent'];
var AnimationPlaybackEvent$1 = window['AnimationPlaybackEvent'];
var ApplicationCache$1 = window['ApplicationCache'];
var Attr$1 = window['Attr'];
var AudioBuffer$1 = window['AudioBuffer'];
var AudioBufferSourceNode$1 = window['AudioBufferSourceNode'];
var AudioContext$1 = window['AudioContext'];
var AudioDestinationNode$1 = window['AudioDestinationNode'];
var AudioListener$1 = window['AudioListener'];
var AudioNode$1 = window['AudioNode'];
var AudioParam$1 = window['AudioParam'];
var AudioProcessingEvent$1 = window['AudioProcessingEvent'];
var AudioTrack$1 = window['AudioTrack'];
var AudioTrackList$1 = window['AudioTrackList'];
var BarProp$1 = window['BarProp'];
var BeforeUnloadEvent$1 = window['BeforeUnloadEvent'];
var BhxBrowser$1 = window['BhxBrowser'];
var BiquadFilterNode$1 = window['BiquadFilterNode'];
var Blob$1 = window['Blob'];
var BroadcastChannel$1 = window['BroadcastChannel'];
var ByteLengthQueuingStrategy$1 = window['ByteLengthQueuingStrategy'];
var CDATASection$1 = window['CDATASection'];
var CSS$1 = window['CSS'];
var CSSConditionRule$1 = window['CSSConditionRule'];
var CSSFontFaceRule$1 = window['CSSFontFaceRule'];
var CSSGroupingRule$1 = window['CSSGroupingRule'];
var CSSImportRule$1 = window['CSSImportRule'];
var CSSKeyframeRule$1 = window['CSSKeyframeRule'];
var CSSKeyframesRule$1 = window['CSSKeyframesRule'];
var CSSMediaRule$1 = window['CSSMediaRule'];
var CSSNamespaceRule$1 = window['CSSNamespaceRule'];
var CSSPageRule$1 = window['CSSPageRule'];
var CSSRule$1 = window['CSSRule'];
var CSSRuleList$1 = window['CSSRuleList'];
var CSSStyleDeclaration$1 = window['CSSStyleDeclaration'];
var CSSStyleRule$1 = window['CSSStyleRule'];
var CSSStyleSheet$1 = window['CSSStyleSheet'];
var CSSSupportsRule$1 = window['CSSSupportsRule'];
var Cache$1 = window['Cache'];
var CacheStorage$1 = window['CacheStorage'];
var CanvasGradient$1 = window['CanvasGradient'];
var CanvasPattern$1 = window['CanvasPattern'];
var CanvasRenderingContext2D$1 = window['CanvasRenderingContext2D'];
var ChannelMergerNode$1 = window['ChannelMergerNode'];
var ChannelSplitterNode$1 = window['ChannelSplitterNode'];
var CharacterData$1 = window['CharacterData'];
var ClientRect$1 = window['ClientRect'];
var ClientRectList$1 = window['ClientRectList'];
var ClipboardEvent$1 = window['ClipboardEvent'];
var CloseEvent$1 = window['CloseEvent'];
var Comment$1 = window['Comment'];
var CompositionEvent$1 = window['CompositionEvent'];
var Console$1 = window['Console'];
var ContentScriptGlobalScope$1 = window['ContentScriptGlobalScope'];
var ConvolverNode$1 = window['ConvolverNode'];
var Coordinates$1 = window['Coordinates'];
var CountQueuingStrategy$1 = window['CountQueuingStrategy'];
var Crypto$1 = window['Crypto'];
var CryptoKey$1 = window['CryptoKey'];
var CryptoKeyPair$1 = window['CryptoKeyPair'];
var CustomEvent$1 = window['CustomEvent'];
var DOMError$1 = window['DOMError'];
var DOMException$1 = window['DOMException'];
var DOMImplementation$1 = window['DOMImplementation'];
var DOMParser$1 = window['DOMParser'];
var DOMRect$1 = window['DOMRect'];
var DOMRectReadOnly$1 = window['DOMRectReadOnly'];
var DOMSettableTokenList$1 = window['DOMSettableTokenList'];
var DOMStringList$1 = window['DOMStringList'];
var DOMStringMap$1 = window['DOMStringMap'];
var DOMTokenList$1 = window['DOMTokenList'];
var DataCue$1 = window['DataCue'];
var DataTransfer$1 = window['DataTransfer'];
var DataTransferItem$1 = window['DataTransferItem'];
var DataTransferItemList$1 = window['DataTransferItemList'];
var DeferredPermissionRequest$1 = window['DeferredPermissionRequest'];
var DelayNode$1 = window['DelayNode'];
var DeviceAcceleration$1 = window['DeviceAcceleration'];
var DeviceLightEvent$1 = window['DeviceLightEvent'];
var DeviceMotionEvent$1 = window['DeviceMotionEvent'];
var DeviceOrientationEvent$1 = window['DeviceOrientationEvent'];
var DeviceRotationRate$1 = window['DeviceRotationRate'];
var Document$1 = window['Document'];
var DocumentFragment$1 = window['DocumentFragment'];
var DocumentType$1 = window['DocumentType'];
var DragEvent$1 = window['DragEvent'];
var DynamicsCompressorNode$1 = window['DynamicsCompressorNode'];
var EXT_texture_filter_anisotropic$1 = window['EXT_texture_filter_anisotropic'];
var Element$1 = window['Element'];
var ErrorEvent$1 = window['ErrorEvent'];
var Event$1 = window['Event'];
var EventSource$1 = window['EventSource'];
var EventTarget$1 = window['EventTarget'];
var ExtensionScriptApis$1 = window['ExtensionScriptApis'];
var External$1 = window['External'];
var File$1 = window['File'];
var FileList$1 = window['FileList'];
var FileReader$1 = window['FileReader'];
var FocusEvent$1 = window['FocusEvent'];
var FocusNavigationEvent$1 = window['FocusNavigationEvent'];
var FormData$1 = window['FormData'];
var GainNode$1 = window['GainNode'];
var Gamepad$1 = window['Gamepad'];
var GamepadButton$1 = window['GamepadButton'];
var GamepadEvent$1 = window['GamepadEvent'];
var GamepadHapticActuator$1 = window['GamepadHapticActuator'];
var GamepadPose$1 = window['GamepadPose'];
var Geolocation$1 = window['Geolocation'];
var HTMLAllCollection$1 = window['HTMLAllCollection'];
var HTMLAnchorElement$1 = window['HTMLAnchorElement'];
var HTMLAppletElement$1 = window['HTMLAppletElement'];
var HTMLAreaElement$1 = window['HTMLAreaElement'];
var HTMLAreasCollection$1 = window['HTMLAreasCollection'];
var HTMLAudioElement$1 = window['HTMLAudioElement'];
var HTMLBRElement$1 = window['HTMLBRElement'];
var HTMLBaseElement$1 = window['HTMLBaseElement'];
var HTMLBaseFontElement$1 = window['HTMLBaseFontElement'];
var HTMLBodyElement$1 = window['HTMLBodyElement'];
var HTMLButtonElement$1 = window['HTMLButtonElement'];
var HTMLCanvasElement$1 = window['HTMLCanvasElement'];
var HTMLCollection$1 = window['HTMLCollection'];
var HTMLDListElement$1 = window['HTMLDListElement'];
var HTMLDataElement$1 = window['HTMLDataElement'];
var HTMLDataListElement$1 = window['HTMLDataListElement'];
var HTMLDetailsElement$1 = window['HTMLDetailsElement'];
var HTMLDialogElement$1 = window['HTMLDialogElement'];
var HTMLDirectoryElement$1 = window['HTMLDirectoryElement'];
var HTMLDivElement$1 = window['HTMLDivElement'];
var HTMLDocument$1 = window['HTMLDocument'];
var HTMLElement$1 = window['HTMLElement'];
var HTMLEmbedElement$1 = window['HTMLEmbedElement'];
var HTMLFieldSetElement$1 = window['HTMLFieldSetElement'];
var HTMLFontElement$1 = window['HTMLFontElement'];
var HTMLFormControlsCollection$1 = window['HTMLFormControlsCollection'];
var HTMLFormElement$1 = window['HTMLFormElement'];
var HTMLFrameElement$1 = window['HTMLFrameElement'];
var HTMLFrameSetElement$1 = window['HTMLFrameSetElement'];
var HTMLHRElement$1 = window['HTMLHRElement'];
var HTMLHeadElement$1 = window['HTMLHeadElement'];
var HTMLHeadingElement$1 = window['HTMLHeadingElement'];
var HTMLHtmlElement$1 = window['HTMLHtmlElement'];
var HTMLIFrameElement$1 = window['HTMLIFrameElement'];
var HTMLImageElement$1 = window['HTMLImageElement'];
var HTMLInputElement$1 = window['HTMLInputElement'];
var HTMLLIElement$1 = window['HTMLLIElement'];
var HTMLLabelElement$1 = window['HTMLLabelElement'];
var HTMLLegendElement$1 = window['HTMLLegendElement'];
var HTMLLinkElement$1 = window['HTMLLinkElement'];
var HTMLMainElement$1 = window['HTMLMainElement'];
var HTMLMapElement$1 = window['HTMLMapElement'];
var HTMLMarqueeElement$1 = window['HTMLMarqueeElement'];
var HTMLMediaElement$1 = window['HTMLMediaElement'];
var HTMLMenuElement$1 = window['HTMLMenuElement'];
var HTMLMetaElement$1 = window['HTMLMetaElement'];
var HTMLMeterElement$1 = window['HTMLMeterElement'];
var HTMLModElement$1 = window['HTMLModElement'];
var HTMLOListElement$1 = window['HTMLOListElement'];
var HTMLObjectElement$1 = window['HTMLObjectElement'];
var HTMLOptGroupElement$1 = window['HTMLOptGroupElement'];
var HTMLOptionElement$1 = window['HTMLOptionElement'];
var HTMLOptionsCollection$1 = window['HTMLOptionsCollection'];
var HTMLOutputElement$1 = window['HTMLOutputElement'];
var HTMLParagraphElement$1 = window['HTMLParagraphElement'];
var HTMLParamElement$1 = window['HTMLParamElement'];
var HTMLPictureElement$1 = window['HTMLPictureElement'];
var HTMLPreElement$1 = window['HTMLPreElement'];
var HTMLProgressElement$1 = window['HTMLProgressElement'];
var HTMLQuoteElement$1 = window['HTMLQuoteElement'];
var HTMLScriptElement$1 = window['HTMLScriptElement'];
var HTMLSelectElement$1 = window['HTMLSelectElement'];
var HTMLSourceElement$1 = window['HTMLSourceElement'];
var HTMLSpanElement$1 = window['HTMLSpanElement'];
var HTMLStyleElement$1 = window['HTMLStyleElement'];
var HTMLSummaryElement$1 = window['HTMLSummaryElement'];
var HTMLTableCaptionElement$1 = window['HTMLTableCaptionElement'];
var HTMLTableCellElement$1 = window['HTMLTableCellElement'];
var HTMLTableColElement$1 = window['HTMLTableColElement'];
var HTMLTableDataCellElement$1 = window['HTMLTableDataCellElement'];
var HTMLTableElement$1 = window['HTMLTableElement'];
var HTMLTableHeaderCellElement$1 = window['HTMLTableHeaderCellElement'];
var HTMLTableRowElement$1 = window['HTMLTableRowElement'];
var HTMLTableSectionElement$1 = window['HTMLTableSectionElement'];
var HTMLTemplateElement$1 = window['HTMLTemplateElement'];
var HTMLTextAreaElement$1 = window['HTMLTextAreaElement'];
var HTMLTimeElement$1 = window['HTMLTimeElement'];
var HTMLTitleElement$1 = window['HTMLTitleElement'];
var HTMLTrackElement$1 = window['HTMLTrackElement'];
var HTMLUListElement$1 = window['HTMLUListElement'];
var HTMLUnknownElement$1 = window['HTMLUnknownElement'];
var HTMLVideoElement$1 = window['HTMLVideoElement'];
var HTMLegendElement$1 = window['HTMLegendElement'];
var HashChangeEvent$1 = window['HashChangeEvent'];
var Headers$1 = window['Headers'];
var History$1 = window['History'];
var IDBCursor$1 = window['IDBCursor'];
var IDBCursorWithValue$1 = window['IDBCursorWithValue'];
var IDBDatabase$1 = window['IDBDatabase'];
var IDBFactory$1 = window['IDBFactory'];
var IDBIndex$1 = window['IDBIndex'];
var IDBKeyRange$1 = window['IDBKeyRange'];
var IDBObjectStore$1 = window['IDBObjectStore'];
var IDBOpenDBRequest$1 = window['IDBOpenDBRequest'];
var IDBRequest$1 = window['IDBRequest'];
var IDBTransaction$1 = window['IDBTransaction'];
var IDBVersionChangeEvent$1 = window['IDBVersionChangeEvent'];
var IIRFilterNode$1 = window['IIRFilterNode'];
var ImageData$1 = window['ImageData'];
var IntersectionObserver$1 = window['IntersectionObserver'];
var IntersectionObserverEntry$1 = window['IntersectionObserverEntry'];
var KeyboardEvent$1 = window['KeyboardEvent'];
var ListeningStateChangedEvent$1 = window['ListeningStateChangedEvent'];
var Location$1 = window['Location'];
var MSAssertion$1 = window['MSAssertion'];
var MSBlobBuilder$1 = window['MSBlobBuilder'];
var MSCredentials$1 = window['MSCredentials'];
var MSDCCEvent$1 = window['MSDCCEvent'];
var MSDSHEvent$1 = window['MSDSHEvent'];
var MSFIDOCredentialAssertion$1 = window['MSFIDOCredentialAssertion'];
var MSFIDOSignature$1 = window['MSFIDOSignature'];
var MSFIDOSignatureAssertion$1 = window['MSFIDOSignatureAssertion'];
var MSGesture$1 = window['MSGesture'];
var MSGestureEvent$1 = window['MSGestureEvent'];
var MSGraphicsTrust$1 = window['MSGraphicsTrust'];
var MSInputMethodContext$1 = window['MSInputMethodContext'];
var MSMediaKeyError$1 = window['MSMediaKeyError'];
var MSMediaKeyMessageEvent$1 = window['MSMediaKeyMessageEvent'];
var MSMediaKeyNeededEvent$1 = window['MSMediaKeyNeededEvent'];
var MSMediaKeySession$1 = window['MSMediaKeySession'];
var MSMediaKeys$1 = window['MSMediaKeys'];
var MSPointerEvent$1 = window['MSPointerEvent'];
var MSStream$1 = window['MSStream'];
var MSStreamReader$1 = window['MSStreamReader'];
var MediaDeviceInfo$1 = window['MediaDeviceInfo'];
var MediaDevices$1 = window['MediaDevices'];
var MediaElementAudioSourceNode$1 = window['MediaElementAudioSourceNode'];
var MediaEncryptedEvent$1 = window['MediaEncryptedEvent'];
var MediaError$1 = window['MediaError'];
var MediaKeyMessageEvent$1 = window['MediaKeyMessageEvent'];
var MediaKeySession$1 = window['MediaKeySession'];
var MediaKeyStatusMap$1 = window['MediaKeyStatusMap'];
var MediaKeySystemAccess$1 = window['MediaKeySystemAccess'];
var MediaKeys$1 = window['MediaKeys'];
var MediaList$1 = window['MediaList'];
var MediaQueryList$1 = window['MediaQueryList'];
var MediaSource$1 = window['MediaSource'];
var MediaStream$1 = window['MediaStream'];
var MediaStreamAudioSourceNode$1 = window['MediaStreamAudioSourceNode'];
var MediaStreamError$1 = window['MediaStreamError'];
var MediaStreamErrorEvent$1 = window['MediaStreamErrorEvent'];
var MediaStreamEvent$1 = window['MediaStreamEvent'];
var MediaStreamTrack$1 = window['MediaStreamTrack'];
var MediaStreamTrackEvent$1 = window['MediaStreamTrackEvent'];
var MessageChannel$1 = window['MessageChannel'];
var MessageEvent$1 = window['MessageEvent'];
var MessagePort$1 = window['MessagePort'];
var MimeType$1 = window['MimeType'];
var MimeTypeArray$1 = window['MimeTypeArray'];
var MouseEvent$1 = window['MouseEvent'];
var MutationEvent$1 = window['MutationEvent'];
var MutationObserver$1 = window['MutationObserver'];
var MutationRecord$1 = window['MutationRecord'];
var NamedNodeMap$1 = window['NamedNodeMap'];
var Navigator$1 = window['Navigator'];
var Node$1 = window['Node'];
var NodeFilter$1 = window['NodeFilter'];
var NodeIterator$1 = window['NodeIterator'];
var NodeList$1 = window['NodeList'];
var Notification$1 = window['Notification'];
var OES_element_index_uint$1 = window['OES_element_index_uint'];
var OES_standard_derivatives$1 = window['OES_standard_derivatives'];
var OES_texture_float$1 = window['OES_texture_float'];
var OES_texture_float_linear$1 = window['OES_texture_float_linear'];
var OES_texture_half_float$1 = window['OES_texture_half_float'];
var OES_texture_half_float_linear$1 = window['OES_texture_half_float_linear'];
var OfflineAudioCompletionEvent$1 = window['OfflineAudioCompletionEvent'];
var OfflineAudioContext$1 = window['OfflineAudioContext'];
var OscillatorNode$1 = window['OscillatorNode'];
var OverflowEvent$1 = window['OverflowEvent'];
var PageTransitionEvent$1 = window['PageTransitionEvent'];
var PannerNode$1 = window['PannerNode'];
var Path2D$1 = window['Path2D'];
var PaymentAddress$1 = window['PaymentAddress'];
var PaymentRequest$1 = window['PaymentRequest'];
var PaymentRequestUpdateEvent$1 = window['PaymentRequestUpdateEvent'];
var PaymentResponse$1 = window['PaymentResponse'];
var PerfWidgetExternal$1 = window['PerfWidgetExternal'];
var Performance$1 = window['Performance'];
var PerformanceEntry$1 = window['PerformanceEntry'];
var PerformanceMark$1 = window['PerformanceMark'];
var PerformanceMeasure$1 = window['PerformanceMeasure'];
var PerformanceNavigation$1 = window['PerformanceNavigation'];
var PerformanceNavigationTiming$1 = window['PerformanceNavigationTiming'];
var PerformanceResourceTiming$1 = window['PerformanceResourceTiming'];
var PerformanceTiming$1 = window['PerformanceTiming'];
var PeriodicWave$1 = window['PeriodicWave'];
var PermissionRequest$1 = window['PermissionRequest'];
var PermissionRequestedEvent$1 = window['PermissionRequestedEvent'];
var Plugin$1 = window['Plugin'];
var PluginArray$1 = window['PluginArray'];
var PointerEvent$1 = window['PointerEvent'];
var PopStateEvent$1 = window['PopStateEvent'];
var Position$1 = window['Position'];
var PositionError$1 = window['PositionError'];
var ProcessingInstruction$1 = window['ProcessingInstruction'];
var ProgressEvent$1 = window['ProgressEvent'];
var PushManager$1 = window['PushManager'];
var PushSubscription$1 = window['PushSubscription'];
var PushSubscriptionOptions$1 = window['PushSubscriptionOptions'];
var RTCDTMFToneChangeEvent$1 = window['RTCDTMFToneChangeEvent'];
var RTCDtlsTransport$1 = window['RTCDtlsTransport'];
var RTCDtlsTransportStateChangedEvent$1 = window['RTCDtlsTransportStateChangedEvent'];
var RTCDtmfSender$1 = window['RTCDtmfSender'];
var RTCIceCandidate$1 = window['RTCIceCandidate'];
var RTCIceCandidatePairChangedEvent$1 = window['RTCIceCandidatePairChangedEvent'];
var RTCIceGatherer$1 = window['RTCIceGatherer'];
var RTCIceGathererEvent$1 = window['RTCIceGathererEvent'];
var RTCIceTransport$1 = window['RTCIceTransport'];
var RTCIceTransportStateChangedEvent$1 = window['RTCIceTransportStateChangedEvent'];
var RTCPeerConnection$1 = window['RTCPeerConnection'];
var RTCPeerConnectionIceEvent$1 = window['RTCPeerConnectionIceEvent'];
var RTCRtpReceiver$1 = window['RTCRtpReceiver'];
var RTCRtpSender$1 = window['RTCRtpSender'];
var RTCSessionDescription$1 = window['RTCSessionDescription'];
var RTCSrtpSdesTransport$1 = window['RTCSrtpSdesTransport'];
var RTCSsrcConflictEvent$1 = window['RTCSsrcConflictEvent'];
var RTCStatsProvider$1 = window['RTCStatsProvider'];
var RandomSource$1 = window['RandomSource'];
var Range$1 = window['Range'];
var ReadableStream$1 = window['ReadableStream'];
var ReadableStreamReader$1 = window['ReadableStreamReader'];
var Request$1 = window['Request'];
var Response$1 = window['Response'];
var SVGAElement$1 = window['SVGAElement'];
var SVGAngle$1 = window['SVGAngle'];
var SVGAnimatedAngle$1 = window['SVGAnimatedAngle'];
var SVGAnimatedBoolean$1 = window['SVGAnimatedBoolean'];
var SVGAnimatedEnumeration$1 = window['SVGAnimatedEnumeration'];
var SVGAnimatedInteger$1 = window['SVGAnimatedInteger'];
var SVGAnimatedLength$1 = window['SVGAnimatedLength'];
var SVGAnimatedLengthList$1 = window['SVGAnimatedLengthList'];
var SVGAnimatedNumber$1 = window['SVGAnimatedNumber'];
var SVGAnimatedNumberList$1 = window['SVGAnimatedNumberList'];
var SVGAnimatedPreserveAspectRatio$1 = window['SVGAnimatedPreserveAspectRatio'];
var SVGAnimatedRect$1 = window['SVGAnimatedRect'];
var SVGAnimatedString$1 = window['SVGAnimatedString'];
var SVGAnimatedTransformList$1 = window['SVGAnimatedTransformList'];
var SVGCircleElement$1 = window['SVGCircleElement'];
var SVGClipPathElement$1 = window['SVGClipPathElement'];
var SVGComponentTransferFunctionElement$1 = window['SVGComponentTransferFunctionElement'];
var SVGDefsElement$1 = window['SVGDefsElement'];
var SVGDescElement$1 = window['SVGDescElement'];
var SVGElement$1 = window['SVGElement'];
var SVGElementInstance$1 = window['SVGElementInstance'];
var SVGElementInstanceList$1 = window['SVGElementInstanceList'];
var SVGEllipseElement$1 = window['SVGEllipseElement'];
var SVGFEBlendElement$1 = window['SVGFEBlendElement'];
var SVGFEColorMatrixElement$1 = window['SVGFEColorMatrixElement'];
var SVGFEComponentTransferElement$1 = window['SVGFEComponentTransferElement'];
var SVGFECompositeElement$1 = window['SVGFECompositeElement'];
var SVGFEConvolveMatrixElement$1 = window['SVGFEConvolveMatrixElement'];
var SVGFEDiffuseLightingElement$1 = window['SVGFEDiffuseLightingElement'];
var SVGFEDisplacementMapElement$1 = window['SVGFEDisplacementMapElement'];
var SVGFEDistantLightElement$1 = window['SVGFEDistantLightElement'];
var SVGFEFloodElement$1 = window['SVGFEFloodElement'];
var SVGFEFuncAElement$1 = window['SVGFEFuncAElement'];
var SVGFEFuncBElement$1 = window['SVGFEFuncBElement'];
var SVGFEFuncGElement$1 = window['SVGFEFuncGElement'];
var SVGFEFuncRElement$1 = window['SVGFEFuncRElement'];
var SVGFEGaussianBlurElement$1 = window['SVGFEGaussianBlurElement'];
var SVGFEImageElement$1 = window['SVGFEImageElement'];
var SVGFEMergeElement$1 = window['SVGFEMergeElement'];
var SVGFEMergeNodeElement$1 = window['SVGFEMergeNodeElement'];
var SVGFEMorphologyElement$1 = window['SVGFEMorphologyElement'];
var SVGFEOffsetElement$1 = window['SVGFEOffsetElement'];
var SVGFEPointLightElement$1 = window['SVGFEPointLightElement'];
var SVGFESpecularLightingElement$1 = window['SVGFESpecularLightingElement'];
var SVGFESpotLightElement$1 = window['SVGFESpotLightElement'];
var SVGFETileElement$1 = window['SVGFETileElement'];
var SVGFETurbulenceElement$1 = window['SVGFETurbulenceElement'];
var SVGFilterElement$1 = window['SVGFilterElement'];
var SVGForeignObjectElement$1 = window['SVGForeignObjectElement'];
var SVGGElement$1 = window['SVGGElement'];
var SVGGradientElement$1 = window['SVGGradientElement'];
var SVGGraphicsElement$1 = window['SVGGraphicsElement'];
var SVGImageElement$1 = window['SVGImageElement'];
var SVGLength$1 = window['SVGLength'];
var SVGLengthList$1 = window['SVGLengthList'];
var SVGLineElement$1 = window['SVGLineElement'];
var SVGLinearGradientElement$1 = window['SVGLinearGradientElement'];
var SVGMarkerElement$1 = window['SVGMarkerElement'];
var SVGMaskElement$1 = window['SVGMaskElement'];
var SVGMatrix$1 = window['SVGMatrix'];
var SVGMetadataElement$1 = window['SVGMetadataElement'];
var SVGNumber$1 = window['SVGNumber'];
var SVGNumberList$1 = window['SVGNumberList'];
var SVGPathElement$1 = window['SVGPathElement'];
var SVGPathSeg$1 = window['SVGPathSeg'];
var SVGPathSegArcAbs$1 = window['SVGPathSegArcAbs'];
var SVGPathSegArcRel$1 = window['SVGPathSegArcRel'];
var SVGPathSegClosePath$1 = window['SVGPathSegClosePath'];
var SVGPathSegCurvetoCubicAbs$1 = window['SVGPathSegCurvetoCubicAbs'];
var SVGPathSegCurvetoCubicRel$1 = window['SVGPathSegCurvetoCubicRel'];
var SVGPathSegCurvetoCubicSmoothAbs$1 = window['SVGPathSegCurvetoCubicSmoothAbs'];
var SVGPathSegCurvetoCubicSmoothRel$1 = window['SVGPathSegCurvetoCubicSmoothRel'];
var SVGPathSegCurvetoQuadraticAbs$1 = window['SVGPathSegCurvetoQuadraticAbs'];
var SVGPathSegCurvetoQuadraticRel$1 = window['SVGPathSegCurvetoQuadraticRel'];
var SVGPathSegCurvetoQuadraticSmoothAbs$1 = window['SVGPathSegCurvetoQuadraticSmoothAbs'];
var SVGPathSegCurvetoQuadraticSmoothRel$1 = window['SVGPathSegCurvetoQuadraticSmoothRel'];
var SVGPathSegLinetoAbs$1 = window['SVGPathSegLinetoAbs'];
var SVGPathSegLinetoHorizontalAbs$1 = window['SVGPathSegLinetoHorizontalAbs'];
var SVGPathSegLinetoHorizontalRel$1 = window['SVGPathSegLinetoHorizontalRel'];
var SVGPathSegLinetoRel$1 = window['SVGPathSegLinetoRel'];
var SVGPathSegLinetoVerticalAbs$1 = window['SVGPathSegLinetoVerticalAbs'];
var SVGPathSegLinetoVerticalRel$1 = window['SVGPathSegLinetoVerticalRel'];
var SVGPathSegList$1 = window['SVGPathSegList'];
var SVGPathSegMovetoAbs$1 = window['SVGPathSegMovetoAbs'];
var SVGPathSegMovetoRel$1 = window['SVGPathSegMovetoRel'];
var SVGPatternElement$1 = window['SVGPatternElement'];
var SVGPoint$1 = window['SVGPoint'];
var SVGPointList$1 = window['SVGPointList'];
var SVGPolygonElement$1 = window['SVGPolygonElement'];
var SVGPolylineElement$1 = window['SVGPolylineElement'];
var SVGPreserveAspectRatio$1 = window['SVGPreserveAspectRatio'];
var SVGRadialGradientElement$1 = window['SVGRadialGradientElement'];
var SVGRect$1 = window['SVGRect'];
var SVGRectElement$1 = window['SVGRectElement'];
var SVGSVGElement$1 = window['SVGSVGElement'];
var SVGScriptElement$1 = window['SVGScriptElement'];
var SVGStopElement$1 = window['SVGStopElement'];
var SVGStringList$1 = window['SVGStringList'];
var SVGStylable$1 = window['SVGStylable'];
var SVGStyleElement$1 = window['SVGStyleElement'];
var SVGSwitchElement$1 = window['SVGSwitchElement'];
var SVGSymbolElement$1 = window['SVGSymbolElement'];
var SVGTSpanElement$1 = window['SVGTSpanElement'];
var SVGTextContentElement$1 = window['SVGTextContentElement'];
var SVGTextElement$1 = window['SVGTextElement'];
var SVGTextPathElement$1 = window['SVGTextPathElement'];
var SVGTextPositioningElement$1 = window['SVGTextPositioningElement'];
var SVGTitleElement$1 = window['SVGTitleElement'];
var SVGTransform$1 = window['SVGTransform'];
var SVGTransformList$1 = window['SVGTransformList'];
var SVGUnitTypes$1 = window['SVGUnitTypes'];
var SVGUseElement$1 = window['SVGUseElement'];
var SVGViewElement$1 = window['SVGViewElement'];
var SVGZoomAndPan$1 = window['SVGZoomAndPan'];
var SVGZoomEvent$1 = window['SVGZoomEvent'];
var ScopedCredential$1 = window['ScopedCredential'];
var ScopedCredentialInfo$1 = window['ScopedCredentialInfo'];
var Screen$1 = window['Screen'];
var ScriptProcessorNode$1 = window['ScriptProcessorNode'];
var SecurityPolicyViolationEvent$1 = window['SecurityPolicyViolationEvent'];
var Selection$1 = window['Selection'];
var ServiceUIFrameContext$1 = window['ServiceUIFrameContext'];
var ServiceWorker$1 = window['ServiceWorker'];
var ServiceWorkerContainer$1 = window['ServiceWorkerContainer'];
var ServiceWorkerMessageEvent$1 = window['ServiceWorkerMessageEvent'];
var ServiceWorkerRegistration$1 = window['ServiceWorkerRegistration'];
var SourceBuffer$1 = window['SourceBuffer'];
var SourceBufferList$1 = window['SourceBufferList'];
var SpeechSynthesis$1 = window['SpeechSynthesis'];
var SpeechSynthesisEvent$1 = window['SpeechSynthesisEvent'];
var SpeechSynthesisUtterance$1 = window['SpeechSynthesisUtterance'];
var SpeechSynthesisVoice$1 = window['SpeechSynthesisVoice'];
var StereoPannerNode$1 = window['StereoPannerNode'];
var Storage$1 = window['Storage'];
var StorageEvent$1 = window['StorageEvent'];
var StyleMedia$1 = window['StyleMedia'];
var StyleSheet$1 = window['StyleSheet'];
var StyleSheetList$1 = window['StyleSheetList'];
var SubtleCrypto$1 = window['SubtleCrypto'];
var SyncManager$1 = window['SyncManager'];
var Text$1 = window['Text'];
var TextDecoder$1 = window['TextDecoder'];
var TextEncoder$1 = window['TextEncoder'];
var TextEvent$1 = window['TextEvent'];
var TextMetrics$1 = window['TextMetrics'];
var TextTrack$1 = window['TextTrack'];
var TextTrackCue$1 = window['TextTrackCue'];
var TextTrackCueList$1 = window['TextTrackCueList'];
var TextTrackList$1 = window['TextTrackList'];
var TimeRanges$1 = window['TimeRanges'];
var Touch$1 = window['Touch'];
var TouchEvent$1 = window['TouchEvent'];
var TouchList$1 = window['TouchList'];
var TrackEvent$1 = window['TrackEvent'];
var TransitionEvent$1 = window['TransitionEvent'];
var TreeWalker$1 = window['TreeWalker'];
var UIEvent$1 = window['UIEvent'];
var URL$1 = window['URL'];
var URLSearchParams$1 = window['URLSearchParams'];
var VRDisplay$1 = window['VRDisplay'];
var VRDisplayCapabilities$1 = window['VRDisplayCapabilities'];
var VRDisplayEvent$1 = window['VRDisplayEvent'];
var VREyeParameters$1 = window['VREyeParameters'];
var VRFieldOfView$1 = window['VRFieldOfView'];
var VRFrameData$1 = window['VRFrameData'];
var VRPose$1 = window['VRPose'];
var ValidityState$1 = window['ValidityState'];
var VideoPlaybackQuality$1 = window['VideoPlaybackQuality'];
var VideoTrack$1 = window['VideoTrack'];
var VideoTrackList$1 = window['VideoTrackList'];
var WEBGL_compressed_texture_s3tc$1 = window['WEBGL_compressed_texture_s3tc'];
var WEBGL_debug_renderer_info$1 = window['WEBGL_debug_renderer_info'];
var WEBGL_depth_texture$1 = window['WEBGL_depth_texture'];
var WaveShaperNode$1 = window['WaveShaperNode'];
var WebAuthentication$1 = window['WebAuthentication'];
var WebAuthnAssertion$1 = window['WebAuthnAssertion'];
var WebGLActiveInfo$1 = window['WebGLActiveInfo'];
var WebGLBuffer$1 = window['WebGLBuffer'];
var WebGLContextEvent$1 = window['WebGLContextEvent'];
var WebGLFramebuffer$1 = window['WebGLFramebuffer'];
var WebGLObject$1 = window['WebGLObject'];
var WebGLProgram$1 = window['WebGLProgram'];
var WebGLRenderbuffer$1 = window['WebGLRenderbuffer'];
var WebGLRenderingContext$1 = window['WebGLRenderingContext'];
var WebGLShader$1 = window['WebGLShader'];
var WebGLShaderPrecisionFormat$1 = window['WebGLShaderPrecisionFormat'];
var WebGLTexture$1 = window['WebGLTexture'];
var WebGLUniformLocation$1 = window['WebGLUniformLocation'];
var WebKitCSSMatrix$1 = window['WebKitCSSMatrix'];
var WebKitDirectoryEntry$1 = window['WebKitDirectoryEntry'];
var WebKitDirectoryReader$1 = window['WebKitDirectoryReader'];
var WebKitEntry$1 = window['WebKitEntry'];
var WebKitFileEntry$1 = window['WebKitFileEntry'];
var WebKitFileSystem$1 = window['WebKitFileSystem'];
var WebKitPoint$1 = window['WebKitPoint'];
var WebSocket$1 = window['WebSocket'];
var WheelEvent$1 = window['WheelEvent'];
var Window$1 = window['Window'];
var Worker$1 = window['Worker'];
var WritableStream$1 = window['WritableStream'];
var WritableStreamDefaultController$1 = window['WritableStreamDefaultController'];
var WritableStreamDefaultWriter$1 = window['WritableStreamDefaultWriter'];
var XMLDocument$1 = window['XMLDocument'];
var XMLHttpRequest$1 = window['XMLHttpRequest'];
var XMLHttpRequestUpload$1 = window['XMLHttpRequestUpload'];
var XMLSerializer$1 = window['XMLSerializer'];
var XPathEvaluator$1 = window['XPathEvaluator'];
var XPathExpression$1 = window['XPathExpression'];
var XPathNSResolver$1 = window['XPathNSResolver'];
var XPathResult$1 = window['XPathResult'];
var XSLTProcessor$1 = window['XSLTProcessor'];
var webkitRTCPeerConnection$1 = window['webkitRTCPeerConnection'];
var Audio$1 = window['Audio'];
var Image$1 = window['Image'];
var Option$1 = window['Option'];
var applicationCache$1 = window['applicationCache'];
var caches$1 = window['caches'];
var clientInformation$1 = window['clientInformation'];
var closed$1 = window['closed'];
var crypto$1 = window['crypto'];
var customElements$1 = window['customElements'];
var defaultStatus$1 = window['defaultStatus'];
var devicePixelRatio$1 = window['devicePixelRatio'];
var doNotTrack$1 = window['doNotTrack'];
var document$1 = window['document'];
var event$1 = window['event'];
var external$1 = window['external'];
var frameElement$1 = window['frameElement'];
var frames$1 = window['frames'];
var history$1 = window['history'];
var innerHeight$1 = window['innerHeight'];
var innerWidth$1 = window['innerWidth'];
var isSecureContext$1 = window['isSecureContext'];
var length$1 = window['length'];
var location$1 = window['location'];
var locationbar$1 = window['locationbar'];
var menubar$1 = window['menubar'];
var msContentScript$1 = window['msContentScript'];
var msCredentials$1 = window['msCredentials'];
var name$1 = window['name'];
var navigator$1 = window['navigator'];
var offscreenBuffering$1 = window['offscreenBuffering'];
var onabort$1 = window['onabort'];
var onafterprint$1 = window['onafterprint'];
var onbeforeprint$1 = window['onbeforeprint'];
var onbeforeunload$1 = window['onbeforeunload'];
var onblur$1 = window['onblur'];
var oncanplay$1 = window['oncanplay'];
var oncanplaythrough$1 = window['oncanplaythrough'];
var onchange$1 = window['onchange'];
var onclick$1 = window['onclick'];
var oncompassneedscalibration$1 = window['oncompassneedscalibration'];
var oncontextmenu$1 = window['oncontextmenu'];
var ondblclick$1 = window['ondblclick'];
var ondevicelight$1 = window['ondevicelight'];
var ondevicemotion$1 = window['ondevicemotion'];
var ondeviceorientation$1 = window['ondeviceorientation'];
var ondrag$1 = window['ondrag'];
var ondragend$1 = window['ondragend'];
var ondragenter$1 = window['ondragenter'];
var ondragleave$1 = window['ondragleave'];
var ondragover$1 = window['ondragover'];
var ondragstart$1 = window['ondragstart'];
var ondrop$1 = window['ondrop'];
var ondurationchange$1 = window['ondurationchange'];
var onemptied$1 = window['onemptied'];
var onended$1 = window['onended'];
var onerror$1 = window['onerror'];
var onfocus$1 = window['onfocus'];
var onhashchange$1 = window['onhashchange'];
var oninput$1 = window['oninput'];
var oninvalid$1 = window['oninvalid'];
var onkeydown$1 = window['onkeydown'];
var onkeypress$1 = window['onkeypress'];
var onkeyup$1 = window['onkeyup'];
var onload$1 = window['onload'];
var onloadeddata$1 = window['onloadeddata'];
var onloadedmetadata$1 = window['onloadedmetadata'];
var onloadstart$1 = window['onloadstart'];
var onmessage$1 = window['onmessage'];
var onmousedown$1 = window['onmousedown'];
var onmouseenter$1 = window['onmouseenter'];
var onmouseleave$1 = window['onmouseleave'];
var onmousemove$1 = window['onmousemove'];
var onmouseout$1 = window['onmouseout'];
var onmouseover$1 = window['onmouseover'];
var onmouseup$1 = window['onmouseup'];
var onmousewheel$1 = window['onmousewheel'];
var onmsgesturechange$1 = window['onmsgesturechange'];
var onmsgesturedoubletap$1 = window['onmsgesturedoubletap'];
var onmsgestureend$1 = window['onmsgestureend'];
var onmsgesturehold$1 = window['onmsgesturehold'];
var onmsgesturestart$1 = window['onmsgesturestart'];
var onmsgesturetap$1 = window['onmsgesturetap'];
var onmsinertiastart$1 = window['onmsinertiastart'];
var onmspointercancel$1 = window['onmspointercancel'];
var onmspointerdown$1 = window['onmspointerdown'];
var onmspointerenter$1 = window['onmspointerenter'];
var onmspointerleave$1 = window['onmspointerleave'];
var onmspointermove$1 = window['onmspointermove'];
var onmspointerout$1 = window['onmspointerout'];
var onmspointerover$1 = window['onmspointerover'];
var onmspointerup$1 = window['onmspointerup'];
var onoffline$1 = window['onoffline'];
var ononline$1 = window['ononline'];
var onorientationchange$1 = window['onorientationchange'];
var onpagehide$1 = window['onpagehide'];
var onpageshow$1 = window['onpageshow'];
var onpause$1 = window['onpause'];
var onplay$1 = window['onplay'];
var onplaying$1 = window['onplaying'];
var onpopstate$1 = window['onpopstate'];
var onprogress$1 = window['onprogress'];
var onratechange$1 = window['onratechange'];
var onreadystatechange$1 = window['onreadystatechange'];
var onreset$1 = window['onreset'];
var onresize$1 = window['onresize'];
var onscroll$1 = window['onscroll'];
var onseeked$1 = window['onseeked'];
var onseeking$1 = window['onseeking'];
var onselect$1 = window['onselect'];
var onstalled$1 = window['onstalled'];
var onstorage$1 = window['onstorage'];
var onsubmit$1 = window['onsubmit'];
var onsuspend$1 = window['onsuspend'];
var ontimeupdate$1 = window['ontimeupdate'];
var ontouchcancel$1 = window['ontouchcancel'];
var ontouchend$1 = window['ontouchend'];
var ontouchmove$1 = window['ontouchmove'];
var ontouchstart$1 = window['ontouchstart'];
var onunload$1 = window['onunload'];
var onvolumechange$1 = window['onvolumechange'];
var onvrdisplayactivate$1 = window['onvrdisplayactivate'];
var onvrdisplayblur$1 = window['onvrdisplayblur'];
var onvrdisplayconnect$1 = window['onvrdisplayconnect'];
var onvrdisplaydeactivate$1 = window['onvrdisplaydeactivate'];
var onvrdisplaydisconnect$1 = window['onvrdisplaydisconnect'];
var onvrdisplayfocus$1 = window['onvrdisplayfocus'];
var onvrdisplaypointerrestricted$1 = window['onvrdisplaypointerrestricted'];
var onvrdisplaypointerunrestricted$1 = window['onvrdisplaypointerunrestricted'];
var onvrdisplaypresentchange$1 = window['onvrdisplaypresentchange'];
var onwaiting$1 = window['onwaiting'];
var opener$1 = window['opener'];
var orientation$1 = window['orientation'];
var outerHeight$1 = window['outerHeight'];
var outerWidth$1 = window['outerWidth'];
var pageXOffset$1 = window['pageXOffset'];
var pageYOffset$1 = window['pageYOffset'];
var parent$1 = window['parent'];
var performance$1 = window['performance'];
var personalbar$1 = window['personalbar'];
var screen$1 = window['screen'];
var screenLeft$1 = window['screenLeft'];
var screenTop$1 = window['screenTop'];
var screenX$1 = window['screenX'];
var screenY$1 = window['screenY'];
var scrollX$1 = window['scrollX'];
var scrollY$1 = window['scrollY'];
var scrollbars$1 = window['scrollbars'];
var self$1 = window['self'];
var speechSynthesis$1 = window['speechSynthesis'];
var status$1 = window['status'];
var statusbar$1 = window['statusbar'];
var styleMedia$1 = window['styleMedia'];
var toolbar$1 = window['toolbar'];
var top$1 = window['top'];
var window$1 = window['window'];
var alert$1 = window['alert'];
var blur$1 = window['blur'];
var cancelAnimationFrame$1 = window['cancelAnimationFrame'];
var captureEvents$1 = window['captureEvents'];
var close$1 = window['close'];
var confirm$1 = window['confirm'];
var createImageBitmap$1 = window['createImageBitmap'];
var departFocus$1 = window['departFocus'];
var focus$1 = window['focus'];
var getComputedStyle$1 = window['getComputedStyle'];
var getMatchedCSSRules$1 = window['getMatchedCSSRules'];
var getSelection$1 = window['getSelection'];
var matchMedia$1 = window['matchMedia'];
var moveBy$1 = window['moveBy'];
var moveTo$1 = window['moveTo'];
var msWriteProfilerMark$1 = window['msWriteProfilerMark'];
var open$1 = window['open'];
var postMessage$1 = window['postMessage'];
var print$1 = window['print'];
var prompt$1 = window['prompt'];
var releaseEvents$1 = window['releaseEvents'];
var requestAnimationFrame$1 = window['requestAnimationFrame'];
var resizeBy$1 = window['resizeBy'];
var resizeTo$1 = window['resizeTo'];
var scroll$1 = window['scroll'];
var scrollBy$1 = window['scrollBy'];
var scrollTo$1 = window['scrollTo'];
var stop$1 = window['stop'];
var webkitCancelAnimationFrame$1 = window['webkitCancelAnimationFrame'];
var webkitConvertPointFromNodeToPage$1 = window['webkitConvertPointFromNodeToPage'];
var webkitConvertPointFromPageToNode$1 = window['webkitConvertPointFromPageToNode'];
var webkitRequestAnimationFrame$1 = window['webkitRequestAnimationFrame'];
var toString$1 = window['toString'];
var dispatchEvent$1 = window['dispatchEvent'];
var clearInterval$1 = window['clearInterval'];
var clearTimeout$1 = window['clearTimeout'];
var setInterval$1 = window['setInterval'];
var setTimeout$1 = window['setTimeout'];
var clearImmediate$1 = window['clearImmediate'];
var setImmediate$1 = window['setImmediate'];
var sessionStorage$1 = window['sessionStorage'];
var localStorage$1 = window['localStorage'];
var console$1 = window['console'];
var onpointercancel$1 = window['onpointercancel'];
var onpointerdown$1 = window['onpointerdown'];
var onpointerenter$1 = window['onpointerenter'];
var onpointerleave$1 = window['onpointerleave'];
var onpointermove$1 = window['onpointermove'];
var onpointerout$1 = window['onpointerout'];
var onpointerover$1 = window['onpointerover'];
var onpointerup$1 = window['onpointerup'];
var onwheel$1 = window['onwheel'];
var indexedDB$1 = window['indexedDB'];
var atob$1 = window['atob'];
var btoa$1 = window['btoa'];
var fetch$1 = window['fetch'];
var addEventListener$1 = window['addEventListener'];
var removeEventListener$1 = window['removeEventListener'];

export {
  ANGLE_instanced_arrays$1 as ANGLE_instanced_arrays,
  AbortController$1 as AbortController,
  AbortSignal$1 as AbortSignal,
  AnalyserNode$1 as AnalyserNode,
  Animation$1 as Animation,
  AnimationEvent$1 as AnimationEvent,
  AnimationPlaybackEvent$1 as AnimationPlaybackEvent,
  ApplicationCache$1 as ApplicationCache,
  Attr$1 as Attr,
  AudioBuffer$1 as AudioBuffer,
  AudioBufferSourceNode$1 as AudioBufferSourceNode,
  AudioContext$1 as AudioContext,
  AudioDestinationNode$1 as AudioDestinationNode,
  AudioListener$1 as AudioListener,
  AudioNode$1 as AudioNode,
  AudioParam$1 as AudioParam,
  AudioProcessingEvent$1 as AudioProcessingEvent,
  AudioTrack$1 as AudioTrack,
  AudioTrackList$1 as AudioTrackList,
  BarProp$1 as BarProp,
  BeforeUnloadEvent$1 as BeforeUnloadEvent,
  BhxBrowser$1 as BhxBrowser,
  BiquadFilterNode$1 as BiquadFilterNode,
  Blob$1 as Blob,
  BroadcastChannel$1 as BroadcastChannel,
  ByteLengthQueuingStrategy$1 as ByteLengthQueuingStrategy,
  CDATASection$1 as CDATASection,
  CSS$1 as CSS,
  CSSConditionRule$1 as CSSConditionRule,
  CSSFontFaceRule$1 as CSSFontFaceRule,
  CSSGroupingRule$1 as CSSGroupingRule,
  CSSImportRule$1 as CSSImportRule,
  CSSKeyframeRule$1 as CSSKeyframeRule,
  CSSKeyframesRule$1 as CSSKeyframesRule,
  CSSMediaRule$1 as CSSMediaRule,
  CSSNamespaceRule$1 as CSSNamespaceRule,
  CSSPageRule$1 as CSSPageRule,
  CSSRule$1 as CSSRule,
  CSSRuleList$1 as CSSRuleList,
  CSSStyleDeclaration$1 as CSSStyleDeclaration,
  CSSStyleRule$1 as CSSStyleRule,
  CSSStyleSheet$1 as CSSStyleSheet,
  CSSSupportsRule$1 as CSSSupportsRule,
  Cache$1 as Cache,
  CacheStorage$1 as CacheStorage,
  CanvasGradient$1 as CanvasGradient,
  CanvasPattern$1 as CanvasPattern,
  CanvasRenderingContext2D$1 as CanvasRenderingContext2D,
  ChannelMergerNode$1 as ChannelMergerNode,
  ChannelSplitterNode$1 as ChannelSplitterNode,
  CharacterData$1 as CharacterData,
  ClientRect$1 as ClientRect,
  ClientRectList$1 as ClientRectList,
  ClipboardEvent$1 as ClipboardEvent,
  CloseEvent$1 as CloseEvent,
  Comment$1 as Comment,
  CompositionEvent$1 as CompositionEvent,
  Console$1 as Console,
  ContentScriptGlobalScope$1 as ContentScriptGlobalScope,
  ConvolverNode$1 as ConvolverNode,
  Coordinates$1 as Coordinates,
  CountQueuingStrategy$1 as CountQueuingStrategy,
  Crypto$1 as Crypto,
  CryptoKey$1 as CryptoKey,
  CryptoKeyPair$1 as CryptoKeyPair,
  CustomEvent$1 as CustomEvent,
  DOMError$1 as DOMError,
  DOMException$1 as DOMException,
  DOMImplementation$1 as DOMImplementation,
  DOMParser$1 as DOMParser,
  DOMRect$1 as DOMRect,
  DOMRectReadOnly$1 as DOMRectReadOnly,
  DOMSettableTokenList$1 as DOMSettableTokenList,
  DOMStringList$1 as DOMStringList,
  DOMStringMap$1 as DOMStringMap,
  DOMTokenList$1 as DOMTokenList,
  DataCue$1 as DataCue,
  DataTransfer$1 as DataTransfer,
  DataTransferItem$1 as DataTransferItem,
  DataTransferItemList$1 as DataTransferItemList,
  DeferredPermissionRequest$1 as DeferredPermissionRequest,
  DelayNode$1 as DelayNode,
  DeviceAcceleration$1 as DeviceAcceleration,
  DeviceLightEvent$1 as DeviceLightEvent,
  DeviceMotionEvent$1 as DeviceMotionEvent,
  DeviceOrientationEvent$1 as DeviceOrientationEvent,
  DeviceRotationRate$1 as DeviceRotationRate,
  Document$1 as Document,
  DocumentFragment$1 as DocumentFragment,
  DocumentType$1 as DocumentType,
  DragEvent$1 as DragEvent,
  DynamicsCompressorNode$1 as DynamicsCompressorNode,
  EXT_texture_filter_anisotropic$1 as EXT_texture_filter_anisotropic,
  Element$1 as Element,
  ErrorEvent$1 as ErrorEvent,
  Event$1 as Event,
  EventSource$1 as EventSource,
  EventTarget$1 as EventTarget,
  ExtensionScriptApis$1 as ExtensionScriptApis,
  External$1 as External,
  File$1 as File,
  FileList$1 as FileList,
  FileReader$1 as FileReader,
  FocusEvent$1 as FocusEvent,
  FocusNavigationEvent$1 as FocusNavigationEvent,
  FormData$1 as FormData,
  GainNode$1 as GainNode,
  Gamepad$1 as Gamepad,
  GamepadButton$1 as GamepadButton,
  GamepadEvent$1 as GamepadEvent,
  GamepadHapticActuator$1 as GamepadHapticActuator,
  GamepadPose$1 as GamepadPose,
  Geolocation$1 as Geolocation,
  HTMLAllCollection$1 as HTMLAllCollection,
  HTMLAnchorElement$1 as HTMLAnchorElement,
  HTMLAppletElement$1 as HTMLAppletElement,
  HTMLAreaElement$1 as HTMLAreaElement,
  HTMLAreasCollection$1 as HTMLAreasCollection,
  HTMLAudioElement$1 as HTMLAudioElement,
  HTMLBRElement$1 as HTMLBRElement,
  HTMLBaseElement$1 as HTMLBaseElement,
  HTMLBaseFontElement$1 as HTMLBaseFontElement,
  HTMLBodyElement$1 as HTMLBodyElement,
  HTMLButtonElement$1 as HTMLButtonElement,
  HTMLCanvasElement$1 as HTMLCanvasElement,
  HTMLCollection$1 as HTMLCollection,
  HTMLDListElement$1 as HTMLDListElement,
  HTMLDataElement$1 as HTMLDataElement,
  HTMLDataListElement$1 as HTMLDataListElement,
  HTMLDetailsElement$1 as HTMLDetailsElement,
  HTMLDialogElement$1 as HTMLDialogElement,
  HTMLDirectoryElement$1 as HTMLDirectoryElement,
  HTMLDivElement$1 as HTMLDivElement,
  HTMLDocument$1 as HTMLDocument,
  HTMLElement$1 as HTMLElement,
  HTMLEmbedElement$1 as HTMLEmbedElement,
  HTMLFieldSetElement$1 as HTMLFieldSetElement,
  HTMLFontElement$1 as HTMLFontElement,
  HTMLFormControlsCollection$1 as HTMLFormControlsCollection,
  HTMLFormElement$1 as HTMLFormElement,
  HTMLFrameElement$1 as HTMLFrameElement,
  HTMLFrameSetElement$1 as HTMLFrameSetElement,
  HTMLHRElement$1 as HTMLHRElement,
  HTMLHeadElement$1 as HTMLHeadElement,
  HTMLHeadingElement$1 as HTMLHeadingElement,
  HTMLHtmlElement$1 as HTMLHtmlElement,
  HTMLIFrameElement$1 as HTMLIFrameElement,
  HTMLImageElement$1 as HTMLImageElement,
  HTMLInputElement$1 as HTMLInputElement,
  HTMLLIElement$1 as HTMLLIElement,
  HTMLLabelElement$1 as HTMLLabelElement,
  HTMLLegendElement$1 as HTMLLegendElement,
  HTMLLinkElement$1 as HTMLLinkElement,
  HTMLMainElement$1 as HTMLMainElement,
  HTMLMapElement$1 as HTMLMapElement,
  HTMLMarqueeElement$1 as HTMLMarqueeElement,
  HTMLMediaElement$1 as HTMLMediaElement,
  HTMLMenuElement$1 as HTMLMenuElement,
  HTMLMetaElement$1 as HTMLMetaElement,
  HTMLMeterElement$1 as HTMLMeterElement,
  HTMLModElement$1 as HTMLModElement,
  HTMLOListElement$1 as HTMLOListElement,
  HTMLObjectElement$1 as HTMLObjectElement,
  HTMLOptGroupElement$1 as HTMLOptGroupElement,
  HTMLOptionElement$1 as HTMLOptionElement,
  HTMLOptionsCollection$1 as HTMLOptionsCollection,
  HTMLOutputElement$1 as HTMLOutputElement,
  HTMLParagraphElement$1 as HTMLParagraphElement,
  HTMLParamElement$1 as HTMLParamElement,
  HTMLPictureElement$1 as HTMLPictureElement,
  HTMLPreElement$1 as HTMLPreElement,
  HTMLProgressElement$1 as HTMLProgressElement,
  HTMLQuoteElement$1 as HTMLQuoteElement,
  HTMLScriptElement$1 as HTMLScriptElement,
  HTMLSelectElement$1 as HTMLSelectElement,
  HTMLSourceElement$1 as HTMLSourceElement,
  HTMLSpanElement$1 as HTMLSpanElement,
  HTMLStyleElement$1 as HTMLStyleElement,
  HTMLSummaryElement$1 as HTMLSummaryElement,
  HTMLTableCaptionElement$1 as HTMLTableCaptionElement,
  HTMLTableCellElement$1 as HTMLTableCellElement,
  HTMLTableColElement$1 as HTMLTableColElement,
  HTMLTableDataCellElement$1 as HTMLTableDataCellElement,
  HTMLTableElement$1 as HTMLTableElement,
  HTMLTableHeaderCellElement$1 as HTMLTableHeaderCellElement,
  HTMLTableRowElement$1 as HTMLTableRowElement,
  HTMLTableSectionElement$1 as HTMLTableSectionElement,
  HTMLTemplateElement$1 as HTMLTemplateElement,
  HTMLTextAreaElement$1 as HTMLTextAreaElement,
  HTMLTimeElement$1 as HTMLTimeElement,
  HTMLTitleElement$1 as HTMLTitleElement,
  HTMLTrackElement$1 as HTMLTrackElement,
  HTMLUListElement$1 as HTMLUListElement,
  HTMLUnknownElement$1 as HTMLUnknownElement,
  HTMLVideoElement$1 as HTMLVideoElement,
  HTMLegendElement$1 as HTMLegendElement,
  HashChangeEvent$1 as HashChangeEvent,
  Headers$1 as Headers,
  History$1 as History,
  IDBCursor$1 as IDBCursor,
  IDBCursorWithValue$1 as IDBCursorWithValue,
  IDBDatabase$1 as IDBDatabase,
  IDBFactory$1 as IDBFactory,
  IDBIndex$1 as IDBIndex,
  IDBKeyRange$1 as IDBKeyRange,
  IDBObjectStore$1 as IDBObjectStore,
  IDBOpenDBRequest$1 as IDBOpenDBRequest,
  IDBRequest$1 as IDBRequest,
  IDBTransaction$1 as IDBTransaction,
  IDBVersionChangeEvent$1 as IDBVersionChangeEvent,
  IIRFilterNode$1 as IIRFilterNode,
  ImageData$1 as ImageData,
  IntersectionObserver$1 as IntersectionObserver,
  IntersectionObserverEntry$1 as IntersectionObserverEntry,
  KeyboardEvent$1 as KeyboardEvent,
  ListeningStateChangedEvent$1 as ListeningStateChangedEvent,
  Location$1 as Location,
  MSAssertion$1 as MSAssertion,
  MSBlobBuilder$1 as MSBlobBuilder,
  MSCredentials$1 as MSCredentials,
  MSDCCEvent$1 as MSDCCEvent,
  MSDSHEvent$1 as MSDSHEvent,
  MSFIDOCredentialAssertion$1 as MSFIDOCredentialAssertion,
  MSFIDOSignature$1 as MSFIDOSignature,
  MSFIDOSignatureAssertion$1 as MSFIDOSignatureAssertion,
  MSGesture$1 as MSGesture,
  MSGestureEvent$1 as MSGestureEvent,
  MSGraphicsTrust$1 as MSGraphicsTrust,
  MSInputMethodContext$1 as MSInputMethodContext,
  MSMediaKeyError$1 as MSMediaKeyError,
  MSMediaKeyMessageEvent$1 as MSMediaKeyMessageEvent,
  MSMediaKeyNeededEvent$1 as MSMediaKeyNeededEvent,
  MSMediaKeySession$1 as MSMediaKeySession,
  MSMediaKeys$1 as MSMediaKeys,
  MSPointerEvent$1 as MSPointerEvent,
  MSStream$1 as MSStream,
  MSStreamReader$1 as MSStreamReader,
  MediaDeviceInfo$1 as MediaDeviceInfo,
  MediaDevices$1 as MediaDevices,
  MediaElementAudioSourceNode$1 as MediaElementAudioSourceNode,
  MediaEncryptedEvent$1 as MediaEncryptedEvent,
  MediaError$1 as MediaError,
  MediaKeyMessageEvent$1 as MediaKeyMessageEvent,
  MediaKeySession$1 as MediaKeySession,
  MediaKeyStatusMap$1 as MediaKeyStatusMap,
  MediaKeySystemAccess$1 as MediaKeySystemAccess,
  MediaKeys$1 as MediaKeys,
  MediaList$1 as MediaList,
  MediaQueryList$1 as MediaQueryList,
  MediaSource$1 as MediaSource,
  MediaStream$1 as MediaStream,
  MediaStreamAudioSourceNode$1 as MediaStreamAudioSourceNode,
  MediaStreamError$1 as MediaStreamError,
  MediaStreamErrorEvent$1 as MediaStreamErrorEvent,
  MediaStreamEvent$1 as MediaStreamEvent,
  MediaStreamTrack$1 as MediaStreamTrack,
  MediaStreamTrackEvent$1 as MediaStreamTrackEvent,
  MessageChannel$1 as MessageChannel,
  MessageEvent$1 as MessageEvent,
  MessagePort$1 as MessagePort,
  MimeType$1 as MimeType,
  MimeTypeArray$1 as MimeTypeArray,
  MouseEvent$1 as MouseEvent,
  MutationEvent$1 as MutationEvent,
  MutationObserver$1 as MutationObserver,
  MutationRecord$1 as MutationRecord,
  NamedNodeMap$1 as NamedNodeMap,
  Navigator$1 as Navigator,
  Node$1 as Node,
  NodeFilter$1 as NodeFilter,
  NodeIterator$1 as NodeIterator,
  NodeList$1 as NodeList,
  Notification$1 as Notification,
  OES_element_index_uint$1 as OES_element_index_uint,
  OES_standard_derivatives$1 as OES_standard_derivatives,
  OES_texture_float$1 as OES_texture_float,
  OES_texture_float_linear$1 as OES_texture_float_linear,
  OES_texture_half_float$1 as OES_texture_half_float,
  OES_texture_half_float_linear$1 as OES_texture_half_float_linear,
  OfflineAudioCompletionEvent$1 as OfflineAudioCompletionEvent,
  OfflineAudioContext$1 as OfflineAudioContext,
  OscillatorNode$1 as OscillatorNode,
  OverflowEvent$1 as OverflowEvent,
  PageTransitionEvent$1 as PageTransitionEvent,
  PannerNode$1 as PannerNode,
  Path2D$1 as Path2D,
  PaymentAddress$1 as PaymentAddress,
  PaymentRequest$1 as PaymentRequest,
  PaymentRequestUpdateEvent$1 as PaymentRequestUpdateEvent,
  PaymentResponse$1 as PaymentResponse,
  PerfWidgetExternal$1 as PerfWidgetExternal,
  Performance$1 as Performance,
  PerformanceEntry$1 as PerformanceEntry,
  PerformanceMark$1 as PerformanceMark,
  PerformanceMeasure$1 as PerformanceMeasure,
  PerformanceNavigation$1 as PerformanceNavigation,
  PerformanceNavigationTiming$1 as PerformanceNavigationTiming,
  PerformanceResourceTiming$1 as PerformanceResourceTiming,
  PerformanceTiming$1 as PerformanceTiming,
  PeriodicWave$1 as PeriodicWave,
  PermissionRequest$1 as PermissionRequest,
  PermissionRequestedEvent$1 as PermissionRequestedEvent,
  Plugin$1 as Plugin,
  PluginArray$1 as PluginArray,
  PointerEvent$1 as PointerEvent,
  PopStateEvent$1 as PopStateEvent,
  Position$1 as Position,
  PositionError$1 as PositionError,
  ProcessingInstruction$1 as ProcessingInstruction,
  ProgressEvent$1 as ProgressEvent,
  PushManager$1 as PushManager,
  PushSubscription$1 as PushSubscription,
  PushSubscriptionOptions$1 as PushSubscriptionOptions,
  RTCDTMFToneChangeEvent$1 as RTCDTMFToneChangeEvent,
  RTCDtlsTransport$1 as RTCDtlsTransport,
  RTCDtlsTransportStateChangedEvent$1 as RTCDtlsTransportStateChangedEvent,
  RTCDtmfSender$1 as RTCDtmfSender,
  RTCIceCandidate$1 as RTCIceCandidate,
  RTCIceCandidatePairChangedEvent$1 as RTCIceCandidatePairChangedEvent,
  RTCIceGatherer$1 as RTCIceGatherer,
  RTCIceGathererEvent$1 as RTCIceGathererEvent,
  RTCIceTransport$1 as RTCIceTransport,
  RTCIceTransportStateChangedEvent$1 as RTCIceTransportStateChangedEvent,
  RTCPeerConnection$1 as RTCPeerConnection,
  RTCPeerConnectionIceEvent$1 as RTCPeerConnectionIceEvent,
  RTCRtpReceiver$1 as RTCRtpReceiver,
  RTCRtpSender$1 as RTCRtpSender,
  RTCSessionDescription$1 as RTCSessionDescription,
  RTCSrtpSdesTransport$1 as RTCSrtpSdesTransport,
  RTCSsrcConflictEvent$1 as RTCSsrcConflictEvent,
  RTCStatsProvider$1 as RTCStatsProvider,
  RandomSource$1 as RandomSource,
  Range$1 as Range,
  ReadableStream$1 as ReadableStream,
  ReadableStreamReader$1 as ReadableStreamReader,
  Request$1 as Request,
  Response$1 as Response,
  SVGAElement$1 as SVGAElement,
  SVGAngle$1 as SVGAngle,
  SVGAnimatedAngle$1 as SVGAnimatedAngle,
  SVGAnimatedBoolean$1 as SVGAnimatedBoolean,
  SVGAnimatedEnumeration$1 as SVGAnimatedEnumeration,
  SVGAnimatedInteger$1 as SVGAnimatedInteger,
  SVGAnimatedLength$1 as SVGAnimatedLength,
  SVGAnimatedLengthList$1 as SVGAnimatedLengthList,
  SVGAnimatedNumber$1 as SVGAnimatedNumber,
  SVGAnimatedNumberList$1 as SVGAnimatedNumberList,
  SVGAnimatedPreserveAspectRatio$1 as SVGAnimatedPreserveAspectRatio,
  SVGAnimatedRect$1 as SVGAnimatedRect,
  SVGAnimatedString$1 as SVGAnimatedString,
  SVGAnimatedTransformList$1 as SVGAnimatedTransformList,
  SVGCircleElement$1 as SVGCircleElement,
  SVGClipPathElement$1 as SVGClipPathElement,
  SVGComponentTransferFunctionElement$1 as SVGComponentTransferFunctionElement,
  SVGDefsElement$1 as SVGDefsElement,
  SVGDescElement$1 as SVGDescElement,
  SVGElement$1 as SVGElement,
  SVGElementInstance$1 as SVGElementInstance,
  SVGElementInstanceList$1 as SVGElementInstanceList,
  SVGEllipseElement$1 as SVGEllipseElement,
  SVGFEBlendElement$1 as SVGFEBlendElement,
  SVGFEColorMatrixElement$1 as SVGFEColorMatrixElement,
  SVGFEComponentTransferElement$1 as SVGFEComponentTransferElement,
  SVGFECompositeElement$1 as SVGFECompositeElement,
  SVGFEConvolveMatrixElement$1 as SVGFEConvolveMatrixElement,
  SVGFEDiffuseLightingElement$1 as SVGFEDiffuseLightingElement,
  SVGFEDisplacementMapElement$1 as SVGFEDisplacementMapElement,
  SVGFEDistantLightElement$1 as SVGFEDistantLightElement,
  SVGFEFloodElement$1 as SVGFEFloodElement,
  SVGFEFuncAElement$1 as SVGFEFuncAElement,
  SVGFEFuncBElement$1 as SVGFEFuncBElement,
  SVGFEFuncGElement$1 as SVGFEFuncGElement,
  SVGFEFuncRElement$1 as SVGFEFuncRElement,
  SVGFEGaussianBlurElement$1 as SVGFEGaussianBlurElement,
  SVGFEImageElement$1 as SVGFEImageElement,
  SVGFEMergeElement$1 as SVGFEMergeElement,
  SVGFEMergeNodeElement$1 as SVGFEMergeNodeElement,
  SVGFEMorphologyElement$1 as SVGFEMorphologyElement,
  SVGFEOffsetElement$1 as SVGFEOffsetElement,
  SVGFEPointLightElement$1 as SVGFEPointLightElement,
  SVGFESpecularLightingElement$1 as SVGFESpecularLightingElement,
  SVGFESpotLightElement$1 as SVGFESpotLightElement,
  SVGFETileElement$1 as SVGFETileElement,
  SVGFETurbulenceElement$1 as SVGFETurbulenceElement,
  SVGFilterElement$1 as SVGFilterElement,
  SVGForeignObjectElement$1 as SVGForeignObjectElement,
  SVGGElement$1 as SVGGElement,
  SVGGradientElement$1 as SVGGradientElement,
  SVGGraphicsElement$1 as SVGGraphicsElement,
  SVGImageElement$1 as SVGImageElement,
  SVGLength$1 as SVGLength,
  SVGLengthList$1 as SVGLengthList,
  SVGLineElement$1 as SVGLineElement,
  SVGLinearGradientElement$1 as SVGLinearGradientElement,
  SVGMarkerElement$1 as SVGMarkerElement,
  SVGMaskElement$1 as SVGMaskElement,
  SVGMatrix$1 as SVGMatrix,
  SVGMetadataElement$1 as SVGMetadataElement,
  SVGNumber$1 as SVGNumber,
  SVGNumberList$1 as SVGNumberList,
  SVGPathElement$1 as SVGPathElement,
  SVGPathSeg$1 as SVGPathSeg,
  SVGPathSegArcAbs$1 as SVGPathSegArcAbs,
  SVGPathSegArcRel$1 as SVGPathSegArcRel,
  SVGPathSegClosePath$1 as SVGPathSegClosePath,
  SVGPathSegCurvetoCubicAbs$1 as SVGPathSegCurvetoCubicAbs,
  SVGPathSegCurvetoCubicRel$1 as SVGPathSegCurvetoCubicRel,
  SVGPathSegCurvetoCubicSmoothAbs$1 as SVGPathSegCurvetoCubicSmoothAbs,
  SVGPathSegCurvetoCubicSmoothRel$1 as SVGPathSegCurvetoCubicSmoothRel,
  SVGPathSegCurvetoQuadraticAbs$1 as SVGPathSegCurvetoQuadraticAbs,
  SVGPathSegCurvetoQuadraticRel$1 as SVGPathSegCurvetoQuadraticRel,
  SVGPathSegCurvetoQuadraticSmoothAbs$1 as SVGPathSegCurvetoQuadraticSmoothAbs,
  SVGPathSegCurvetoQuadraticSmoothRel$1 as SVGPathSegCurvetoQuadraticSmoothRel,
  SVGPathSegLinetoAbs$1 as SVGPathSegLinetoAbs,
  SVGPathSegLinetoHorizontalAbs$1 as SVGPathSegLinetoHorizontalAbs,
  SVGPathSegLinetoHorizontalRel$1 as SVGPathSegLinetoHorizontalRel,
  SVGPathSegLinetoRel$1 as SVGPathSegLinetoRel,
  SVGPathSegLinetoVerticalAbs$1 as SVGPathSegLinetoVerticalAbs,
  SVGPathSegLinetoVerticalRel$1 as SVGPathSegLinetoVerticalRel,
  SVGPathSegList$1 as SVGPathSegList,
  SVGPathSegMovetoAbs$1 as SVGPathSegMovetoAbs,
  SVGPathSegMovetoRel$1 as SVGPathSegMovetoRel,
  SVGPatternElement$1 as SVGPatternElement,
  SVGPoint$1 as SVGPoint,
  SVGPointList$1 as SVGPointList,
  SVGPolygonElement$1 as SVGPolygonElement,
  SVGPolylineElement$1 as SVGPolylineElement,
  SVGPreserveAspectRatio$1 as SVGPreserveAspectRatio,
  SVGRadialGradientElement$1 as SVGRadialGradientElement,
  SVGRect$1 as SVGRect,
  SVGRectElement$1 as SVGRectElement,
  SVGSVGElement$1 as SVGSVGElement,
  SVGScriptElement$1 as SVGScriptElement,
  SVGStopElement$1 as SVGStopElement,
  SVGStringList$1 as SVGStringList,
  SVGStylable$1 as SVGStylable,
  SVGStyleElement$1 as SVGStyleElement,
  SVGSwitchElement$1 as SVGSwitchElement,
  SVGSymbolElement$1 as SVGSymbolElement,
  SVGTSpanElement$1 as SVGTSpanElement,
  SVGTextContentElement$1 as SVGTextContentElement,
  SVGTextElement$1 as SVGTextElement,
  SVGTextPathElement$1 as SVGTextPathElement,
  SVGTextPositioningElement$1 as SVGTextPositioningElement,
  SVGTitleElement$1 as SVGTitleElement,
  SVGTransform$1 as SVGTransform,
  SVGTransformList$1 as SVGTransformList,
  SVGUnitTypes$1 as SVGUnitTypes,
  SVGUseElement$1 as SVGUseElement,
  SVGViewElement$1 as SVGViewElement,
  SVGZoomAndPan$1 as SVGZoomAndPan,
  SVGZoomEvent$1 as SVGZoomEvent,
  ScopedCredential$1 as ScopedCredential,
  ScopedCredentialInfo$1 as ScopedCredentialInfo,
  Screen$1 as Screen,
  ScriptProcessorNode$1 as ScriptProcessorNode,
  SecurityPolicyViolationEvent$1 as SecurityPolicyViolationEvent,
  Selection$1 as Selection,
  ServiceUIFrameContext$1 as ServiceUIFrameContext,
  ServiceWorker$1 as ServiceWorker,
  ServiceWorkerContainer$1 as ServiceWorkerContainer,
  ServiceWorkerMessageEvent$1 as ServiceWorkerMessageEvent,
  ServiceWorkerRegistration$1 as ServiceWorkerRegistration,
  SourceBuffer$1 as SourceBuffer,
  SourceBufferList$1 as SourceBufferList,
  SpeechSynthesis$1 as SpeechSynthesis,
  SpeechSynthesisEvent$1 as SpeechSynthesisEvent,
  SpeechSynthesisUtterance$1 as SpeechSynthesisUtterance,
  SpeechSynthesisVoice$1 as SpeechSynthesisVoice,
  StereoPannerNode$1 as StereoPannerNode,
  Storage$1 as Storage,
  StorageEvent$1 as StorageEvent,
  StyleMedia$1 as StyleMedia,
  StyleSheet$1 as StyleSheet,
  StyleSheetList$1 as StyleSheetList,
  SubtleCrypto$1 as SubtleCrypto,
  SyncManager$1 as SyncManager,
  Text$1 as Text,
  TextDecoder$1 as TextDecoder,
  TextEncoder$1 as TextEncoder,
  TextEvent$1 as TextEvent,
  TextMetrics$1 as TextMetrics,
  TextTrack$1 as TextTrack,
  TextTrackCue$1 as TextTrackCue,
  TextTrackCueList$1 as TextTrackCueList,
  TextTrackList$1 as TextTrackList,
  TimeRanges$1 as TimeRanges,
  Touch$1 as Touch,
  TouchEvent$1 as TouchEvent,
  TouchList$1 as TouchList,
  TrackEvent$1 as TrackEvent,
  TransitionEvent$1 as TransitionEvent,
  TreeWalker$1 as TreeWalker,
  UIEvent$1 as UIEvent,
  URL$1 as URL,
  URLSearchParams$1 as URLSearchParams,
  VRDisplay$1 as VRDisplay,
  VRDisplayCapabilities$1 as VRDisplayCapabilities,
  VRDisplayEvent$1 as VRDisplayEvent,
  VREyeParameters$1 as VREyeParameters,
  VRFieldOfView$1 as VRFieldOfView,
  VRFrameData$1 as VRFrameData,
  VRPose$1 as VRPose,
  ValidityState$1 as ValidityState,
  VideoPlaybackQuality$1 as VideoPlaybackQuality,
  VideoTrack$1 as VideoTrack,
  VideoTrackList$1 as VideoTrackList,
  WEBGL_compressed_texture_s3tc$1 as WEBGL_compressed_texture_s3tc,
  WEBGL_debug_renderer_info$1 as WEBGL_debug_renderer_info,
  WEBGL_depth_texture$1 as WEBGL_depth_texture,
  WaveShaperNode$1 as WaveShaperNode,
  WebAuthentication$1 as WebAuthentication,
  WebAuthnAssertion$1 as WebAuthnAssertion,
  WebGLActiveInfo$1 as WebGLActiveInfo,
  WebGLBuffer$1 as WebGLBuffer,
  WebGLContextEvent$1 as WebGLContextEvent,
  WebGLFramebuffer$1 as WebGLFramebuffer,
  WebGLObject$1 as WebGLObject,
  WebGLProgram$1 as WebGLProgram,
  WebGLRenderbuffer$1 as WebGLRenderbuffer,
  WebGLRenderingContext$1 as WebGLRenderingContext,
  WebGLShader$1 as WebGLShader,
  WebGLShaderPrecisionFormat$1 as WebGLShaderPrecisionFormat,
  WebGLTexture$1 as WebGLTexture,
  WebGLUniformLocation$1 as WebGLUniformLocation,
  WebKitCSSMatrix$1 as WebKitCSSMatrix,
  WebKitDirectoryEntry$1 as WebKitDirectoryEntry,
  WebKitDirectoryReader$1 as WebKitDirectoryReader,
  WebKitEntry$1 as WebKitEntry,
  WebKitFileEntry$1 as WebKitFileEntry,
  WebKitFileSystem$1 as WebKitFileSystem,
  WebKitPoint$1 as WebKitPoint,
  WebSocket$1 as WebSocket,
  WheelEvent$1 as WheelEvent,
  Window$1 as Window,
  Worker$1 as Worker,
  WritableStream$1 as WritableStream,
  WritableStreamDefaultController$1 as WritableStreamDefaultController,
  WritableStreamDefaultWriter$1 as WritableStreamDefaultWriter,
  XMLDocument$1 as XMLDocument,
  XMLHttpRequest$1 as XMLHttpRequest,
  XMLHttpRequestUpload$1 as XMLHttpRequestUpload,
  XMLSerializer$1 as XMLSerializer,
  XPathEvaluator$1 as XPathEvaluator,
  XPathExpression$1 as XPathExpression,
  XPathNSResolver$1 as XPathNSResolver,
  XPathResult$1 as XPathResult,
  XSLTProcessor$1 as XSLTProcessor,
  webkitRTCPeerConnection$1 as webkitRTCPeerConnection,
  Audio$1 as Audio,
  Image$1 as Image,
  Option$1 as Option,
  applicationCache$1 as applicationCache,
  caches$1 as caches,
  clientInformation$1 as clientInformation,
  closed$1 as closed,
  crypto$1 as crypto,
  customElements$1 as customElements,
  defaultStatus$1 as defaultStatus,
  devicePixelRatio$1 as devicePixelRatio,
  doNotTrack$1 as doNotTrack,
  document$1 as document,
  event$1 as event,
  external$1 as external,
  frameElement$1 as frameElement,
  frames$1 as frames,
  history$1 as history,
  innerHeight$1 as innerHeight,
  innerWidth$1 as innerWidth,
  isSecureContext$1 as isSecureContext,
  length$1 as length,
  location$1 as location,
  locationbar$1 as locationbar,
  menubar$1 as menubar,
  msContentScript$1 as msContentScript,
  msCredentials$1 as msCredentials,
  name$1 as name,
  navigator$1 as navigator,
  offscreenBuffering$1 as offscreenBuffering,
  onabort$1 as onabort,
  onafterprint$1 as onafterprint,
  onbeforeprint$1 as onbeforeprint,
  onbeforeunload$1 as onbeforeunload,
  onblur$1 as onblur,
  oncanplay$1 as oncanplay,
  oncanplaythrough$1 as oncanplaythrough,
  onchange$1 as onchange,
  onclick$1 as onclick,
  oncompassneedscalibration$1 as oncompassneedscalibration,
  oncontextmenu$1 as oncontextmenu,
  ondblclick$1 as ondblclick,
  ondevicelight$1 as ondevicelight,
  ondevicemotion$1 as ondevicemotion,
  ondeviceorientation$1 as ondeviceorientation,
  ondrag$1 as ondrag,
  ondragend$1 as ondragend,
  ondragenter$1 as ondragenter,
  ondragleave$1 as ondragleave,
  ondragover$1 as ondragover,
  ondragstart$1 as ondragstart,
  ondrop$1 as ondrop,
  ondurationchange$1 as ondurationchange,
  onemptied$1 as onemptied,
  onended$1 as onended,
  onerror$1 as onerror,
  onfocus$1 as onfocus,
  onhashchange$1 as onhashchange,
  oninput$1 as oninput,
  oninvalid$1 as oninvalid,
  onkeydown$1 as onkeydown,
  onkeypress$1 as onkeypress,
  onkeyup$1 as onkeyup,
  onload$1 as onload,
  onloadeddata$1 as onloadeddata,
  onloadedmetadata$1 as onloadedmetadata,
  onloadstart$1 as onloadstart,
  onmessage$1 as onmessage,
  onmousedown$1 as onmousedown,
  onmouseenter$1 as onmouseenter,
  onmouseleave$1 as onmouseleave,
  onmousemove$1 as onmousemove,
  onmouseout$1 as onmouseout,
  onmouseover$1 as onmouseover,
  onmouseup$1 as onmouseup,
  onmousewheel$1 as onmousewheel,
  onmsgesturechange$1 as onmsgesturechange,
  onmsgesturedoubletap$1 as onmsgesturedoubletap,
  onmsgestureend$1 as onmsgestureend,
  onmsgesturehold$1 as onmsgesturehold,
  onmsgesturestart$1 as onmsgesturestart,
  onmsgesturetap$1 as onmsgesturetap,
  onmsinertiastart$1 as onmsinertiastart,
  onmspointercancel$1 as onmspointercancel,
  onmspointerdown$1 as onmspointerdown,
  onmspointerenter$1 as onmspointerenter,
  onmspointerleave$1 as onmspointerleave,
  onmspointermove$1 as onmspointermove,
  onmspointerout$1 as onmspointerout,
  onmspointerover$1 as onmspointerover,
  onmspointerup$1 as onmspointerup,
  onoffline$1 as onoffline,
  ononline$1 as ononline,
  onorientationchange$1 as onorientationchange,
  onpagehide$1 as onpagehide,
  onpageshow$1 as onpageshow,
  onpause$1 as onpause,
  onplay$1 as onplay,
  onplaying$1 as onplaying,
  onpopstate$1 as onpopstate,
  onprogress$1 as onprogress,
  onratechange$1 as onratechange,
  onreadystatechange$1 as onreadystatechange,
  onreset$1 as onreset,
  onresize$1 as onresize,
  onscroll$1 as onscroll,
  onseeked$1 as onseeked,
  onseeking$1 as onseeking,
  onselect$1 as onselect,
  onstalled$1 as onstalled,
  onstorage$1 as onstorage,
  onsubmit$1 as onsubmit,
  onsuspend$1 as onsuspend,
  ontimeupdate$1 as ontimeupdate,
  ontouchcancel$1 as ontouchcancel,
  ontouchend$1 as ontouchend,
  ontouchmove$1 as ontouchmove,
  ontouchstart$1 as ontouchstart,
  onunload$1 as onunload,
  onvolumechange$1 as onvolumechange,
  onvrdisplayactivate$1 as onvrdisplayactivate,
  onvrdisplayblur$1 as onvrdisplayblur,
  onvrdisplayconnect$1 as onvrdisplayconnect,
  onvrdisplaydeactivate$1 as onvrdisplaydeactivate,
  onvrdisplaydisconnect$1 as onvrdisplaydisconnect,
  onvrdisplayfocus$1 as onvrdisplayfocus,
  onvrdisplaypointerrestricted$1 as onvrdisplaypointerrestricted,
  onvrdisplaypointerunrestricted$1 as onvrdisplaypointerunrestricted,
  onvrdisplaypresentchange$1 as onvrdisplaypresentchange,
  onwaiting$1 as onwaiting,
  opener$1 as opener,
  orientation$1 as orientation,
  outerHeight$1 as outerHeight,
  outerWidth$1 as outerWidth,
  pageXOffset$1 as pageXOffset,
  pageYOffset$1 as pageYOffset,
  parent$1 as parent,
  performance$1 as performance,
  personalbar$1 as personalbar,
  screen$1 as screen,
  screenLeft$1 as screenLeft,
  screenTop$1 as screenTop,
  screenX$1 as screenX,
  screenY$1 as screenY,
  scrollX$1 as scrollX,
  scrollY$1 as scrollY,
  scrollbars$1 as scrollbars,
  self$1 as self,
  speechSynthesis$1 as speechSynthesis,
  status$1 as status,
  statusbar$1 as statusbar,
  styleMedia$1 as styleMedia,
  toolbar$1 as toolbar,
  top$1 as top,
  window$1 as window,
  alert$1 as alert,
  blur$1 as blur,
  cancelAnimationFrame$1 as cancelAnimationFrame,
  captureEvents$1 as captureEvents,
  close$1 as close,
  confirm$1 as confirm,
  createImageBitmap$1 as createImageBitmap,
  departFocus$1 as departFocus,
  focus$1 as focus,
  getComputedStyle$1 as getComputedStyle,
  getMatchedCSSRules$1 as getMatchedCSSRules,
  getSelection$1 as getSelection,
  matchMedia$1 as matchMedia,
  moveBy$1 as moveBy,
  moveTo$1 as moveTo,
  msWriteProfilerMark$1 as msWriteProfilerMark,
  open$1 as open,
  postMessage$1 as postMessage,
  print$1 as print,
  prompt$1 as prompt,
  releaseEvents$1 as releaseEvents,
  requestAnimationFrame$1 as requestAnimationFrame,
  resizeBy$1 as resizeBy,
  resizeTo$1 as resizeTo,
  scroll$1 as scroll,
  scrollBy$1 as scrollBy,
  scrollTo$1 as scrollTo,
  stop$1 as stop,
  webkitCancelAnimationFrame$1 as webkitCancelAnimationFrame,
  webkitConvertPointFromNodeToPage$1 as webkitConvertPointFromNodeToPage,
  webkitConvertPointFromPageToNode$1 as webkitConvertPointFromPageToNode,
  webkitRequestAnimationFrame$1 as webkitRequestAnimationFrame,
  toString$1 as toString,
  dispatchEvent$1 as dispatchEvent,
  clearInterval$1 as clearInterval,
  clearTimeout$1 as clearTimeout,
  setInterval$1 as setInterval,
  setTimeout$1 as setTimeout,
  clearImmediate$1 as clearImmediate,
  setImmediate$1 as setImmediate,
  sessionStorage$1 as sessionStorage,
  localStorage$1 as localStorage,
  console$1 as console,
  onpointercancel$1 as onpointercancel,
  onpointerdown$1 as onpointerdown,
  onpointerenter$1 as onpointerenter,
  onpointerleave$1 as onpointerleave,
  onpointermove$1 as onpointermove,
  onpointerout$1 as onpointerout,
  onpointerover$1 as onpointerover,
  onpointerup$1 as onpointerup,
  onwheel$1 as onwheel,
  indexedDB$1 as indexedDB,
  atob$1 as atob,
  btoa$1 as btoa,
  fetch$1 as fetch,
  addEventListener$1 as addEventListener,
  removeEventListener$1 as removeEventListener
}
