// @ephox/katamari/lib/main/ts/ephox/katamari/api/Option.js
import * as Fun from './Fun';
var none = function () { return NONE; };
var NONE = (function () {
    var eq = function (o) {
        return o.isNone();
    };
    // inlined from peanut, maybe a micro-optimisation?
    var call = function (thunk) { return thunk(); };
    var id = function (n) { return n; };
    var me = {
        fold: function (n, s) { return n(); },
        is: Fun.never,
        isSome: Fun.never,
        isNone: Fun.always,
        getOr: id,
        getOrThunk: call,
        getOrDie: function (msg) {
            throw new Error(msg || 'error: getOrDie called on none.');
        },
        getOrNull: Fun.constant(null),
        getOrUndefined: Fun.constant(undefined),
        or: id,
        orThunk: call,
        map: none,
        each: Fun.noop,
        bind: none,
        exists: Fun.never,
        forall: Fun.always,
        filter: none,
        equals: eq,
        equals_: eq,
        toArray: function () { return []; },
        toString: Fun.constant('none()')
    };
    if (Object.freeze) {
        Object.freeze(me);
    }
    return me;
})();
var some = function (a) {
    var constant_a = Fun.constant(a);
    var self = function () {
        // can't Fun.constant this one
        return me;
    };
    var bind = function (f) {
        return f(a);
    };
    var me = {
        fold: function (n, s) { return s(a); },
        is: function (v) { return a === v; },
        isSome: Fun.always,
        isNone: Fun.never,
        getOr: constant_a,
        getOrThunk: constant_a,
        getOrDie: constant_a,
        getOrNull: constant_a,
        getOrUndefined: constant_a,
        or: self,
        orThunk: self,
        map: function (f) { return some(f(a)); },
        each: function (f) {
            f(a);
        },
        bind: bind,
        exists: bind,
        forall: bind,
        filter: function (f) {
            return f(a) ? me : NONE;
        },
        toArray: function () { return [a]; },
        toString: function () { return 'some(' + a + ')'; },
        equals: function (o) {
            return o.is(a);
        },
        equals_: function (o, elementEq) {
            return o.fold(Fun.never, function (b) { return elementEq(a, b); });
        },
    };
    return me;
};
var from = function (value) { return value === null || value === undefined ? NONE : some(value); };
export var Option = {
    some: some,
    none: none,
    from: from
};
