// @ephox/katamari/lib/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.js
import { Option } from 'ephox/katamari/api/Option';
import { Result } from 'ephox/katamari/api/Result';
import { Future } from 'ephox/katamari/api/Future';
import { setTimeout } from '@ephox/dom-globals';
import fc from 'fast-check';
export var arbResultError = function (arbE) {
    return arbE.map(Result.error);
};
export var arbResultValue = function (arbA) {
    return arbA.map(Result.value);
};
export var arbResult = function (arbA, arbE) {
    return fc.oneof(arbResultError(arbE), arbResultValue(arbA));
};
export var arbOptionNone = function () { return fc.constant(Option.none()); };
export var arbOptionSome = function (at) { return at.map(Option.some); };
export var arbOption = function (at) { return fc.oneof(arbOptionNone(), arbOptionSome(at)); };
export var arbNegativeInteger = function () { return fc.integer(Number.MIN_SAFE_INTEGER, -1); };
export var arbFutureNow = function (arbA) {
    return arbA.map(Future.pure);
};
export var arbFutureSoon = function (arbA) {
    return arbA.map(function (a) { return Future.nu(function (cb) {
        setTimeout(function () {
            cb(a);
        }, 5);
    }); });
};
export var arbFutureNever = function () {
    return fc.constant(Future.nu(function () { }));
};
export var arbFutureNowOrSoon = function (arbA) {
    return fc.oneof(arbFutureNow(arbA), arbFutureSoon(arbA));
};
