// @ephox/agar/lib/main/ts/ephox/agar/api/Mouse.js
import { Focus } from '@ephox/sugar';
import * as Clicks from '../mouse/Clicks';
import { Chain } from './Chain';
import * as UiFinder from './UiFinder';
var cTrigger = function (selector, action) {
    return Chain.async(function (container, next, die) {
        UiFinder.findIn(container, selector).fold(function () { return die('Could not find element: ' + selector); }, function (ele) {
            action(ele);
            next(container);
        });
    });
};
var sTriggerWith = function (container, selector, action) {
    return Chain.asStep(container, [cTrigger(selector, action)]);
};
var trueClick = function (elem) {
    // The closest event queue to a true Click
    Focus.focus(elem);
    Clicks.mousedown(elem);
    Clicks.mouseup(elem);
    Clicks.trigger(elem);
};
var sClickOn = function (container, selector) {
    return sTriggerWith(container, selector, Clicks.trigger);
};
var sHoverOn = function (container, selector) {
    return sTriggerWith(container, selector, Clicks.mouseover);
};
var sTrueClickOn = function (container, selector) {
    return sTriggerWith(container, selector, trueClick);
};
var sContextMenuOn = function (container, selector) {
    return sTriggerWith(container, selector, Clicks.contextmenu);
};
var cClickOn = function (selector) {
    return cTrigger(selector, Clicks.trigger);
};
var cMouseUpTo = function (dx, dy) {
    return Chain.op(Clicks.mouseupTo(dx, dy));
};
var cMouseMoveTo = function (dx, dy) {
    return Chain.op(Clicks.mousemoveTo(dx, dy));
};
var point = Clicks.point;
var cClick = Chain.op(Clicks.trigger);
var cTrueClick = Chain.op(trueClick);
var cContextMenu = Chain.op(Clicks.contextmenu);
var cMouseOver = Chain.op(Clicks.mouseover);
var cMouseDown = Chain.op(Clicks.mousedown);
var cMouseUp = Chain.op(Clicks.mouseup);
var cMouseMove = Chain.op(Clicks.mousemove);
var cMouseOut = Chain.op(Clicks.mouseout);
export { point, sClickOn, sTrueClickOn, sHoverOn, sContextMenuOn, cClick, cClickOn, cTrueClick, cContextMenu, cMouseOver, cMouseDown, cMouseUp, cMouseUpTo, cMouseMove, cMouseMoveTo, cMouseOut };
