// @ephox/agar/lib/main/ts/ephox/agar/api/RealMouse.js
import * as SeleniumAction from '../server/SeleniumAction';
import { Chain } from './Chain';
import { Id, Fun } from '@ephox/katamari';
import { Attr } from '@ephox/sugar';
var BedrockIdAttribute = 'data-bedrockid';
var sActionOn = function (selector, type) {
    return SeleniumAction.sPerform('/mouse', {
        selector: selector,
        type: type
    });
};
var sMoveToOn = function (selector) {
    return sActionOn(selector, 'move');
};
var sDownOn = function (selector) {
    return sActionOn(selector, 'down');
};
var sUpOn = function (selector) {
    return sActionOn(selector, 'up');
};
var sClickOn = function (selector) {
    return sActionOn(selector, 'click');
};
var cAction = function (action) {
    return Chain.fromChains([
        Chain.mapper(function (selector) { return ({
            selector: selector,
            type: action
        }); }),
        SeleniumAction.cPerform('/mouse')
    ]);
};
var cClick = function () {
    return Chain.fromParent(Chain.mapper(Fun.identity), [
        Chain.fromChains([
            Chain.mapper(function (elem) {
                var id = Id.generate('');
                Attr.set(elem, BedrockIdAttribute, id);
                return "[" + BedrockIdAttribute + "=\"" + id + "\"]";
            }),
            cAction('click')
        ]),
        Chain.op(function (elem) {
            Attr.remove(elem, BedrockIdAttribute);
        })
    ]);
};
export { sMoveToOn, sDownOn, sUpOn, sClickOn, cClick, BedrockIdAttribute };
